File: backend/app/coach/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class CoachConfigRequest(BaseModel):
    provider: str
    model: str
    # Blank on update means "keep the stored key" (mirrors the sync config).
    api_key: str = ""
    base_url: str | None = None
    enabled: bool = True


class CoachConfigResponse(BaseModel):
    provider: str
    model: str
    has_api_key: bool
    base_url: str | None
    enabled: bool
    last_status: str | None
    last_message: str | None
    updated_on: datetime | None = None


class CoachVerifyRequest(BaseModel):
    """Settings to test. Any field left unset falls back to the stored config."""

    provider: str | None = None
    model: str | None = None
    api_key: str = ""
    base_url: str | None = None


class CoachVerifyResult(BaseModel):
    ok: bool
    message: str


class CoachStatusResponse(BaseModel):
    configured: bool
    enabled: bool
    # usable = configured and enabled and the last verification succeeded.
    # The frontend shows the coach launcher only when this is true.
    usable: bool
    provider: str | None = None
    model: str | None = None
    last_status: str | None = None
    last_message: str | None = None


class CoachSessionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    created_on: datetime
    updated_on: datetime


class CoachMessageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    role: str
    content: str
    created_on: datetime


class CoachSessionRenameRequest(BaseModel):
    title: str


class CoachChatContext(BaseModel):
    """What the user is currently viewing, derived from the route by the UI."""

    view: str | None = None
    activity_id: str | None = None


class CoachSkillResponse(BaseModel):
    """A selectable coaching skill, for the chat "/" menu. Body is not exposed."""

    id: str
    name: str
    description: str


class CoachChatRequest(BaseModel):
    message: str
    # When omitted a new session is created and its id returned in the done event.
    session_id: int | None = None
    context: CoachChatContext | None = None
    # Optional skill id chosen from the "/" menu; applied to this message only.
    skill: str | None = None


class CoachMemoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    content: str
    created_on: datetime


class PlannedSession(BaseModel):
    day: str
    sport: str
    workout_type: str
    description: str
    target_distance_km: float | None = None
    target_duration_min: int | None = None
    intensity: str | None = None


class PlannedWeek(BaseModel):
    week: int
    focus: str
    sessions: list[PlannedSession]


class TrainingPlan(BaseModel):
    title: str
    summary: str
    weeks: list[PlannedWeek]


class CoachPlanRequest(BaseModel):
    goal: str
    weeks: int = 4
    context: CoachChatContext | None = None


class CoachPlanResponse(BaseModel):
    # Exactly one is set: a structured plan, or a clarifying message from the coach.
    plan: TrainingPlan | None = None
    message: str | None = None

File: frontend/lib/coach/schemas.ts
import { z } from "zod"

export const CoachStatusSchema = z.object({
  configured: z.boolean(),
  enabled: z.boolean(),
  usable: z.boolean(),
  provider: z.string().nullable(),
  model: z.string().nullable(),
  last_status: z.string().nullable(),
  last_message: z.string().nullable(),
})

export const CoachConfigSchema = z.object({
  provider: z.string(),
  model: z.string(),
  has_api_key: z.boolean(),
  base_url: z.string().nullable(),
  enabled: z.boolean(),
  last_status: z.string().nullable(),
  last_message: z.string().nullable(),
  updated_on: z.string().nullable().optional(),
})

export const CoachVerifyResultSchema = z.object({
  ok: z.boolean(),
  message: z.string(),
})

export const CoachSessionSchema = z.object({
  id: z.number(),
  title: z.string(),
  created_on: z.string(),
  updated_on: z.string(),
})

export const CoachMessageSchema = z.object({
  id: z.number(),
  role: z.enum(["user", "assistant"]),
  content: z.string(),
  created_on: z.string(),
})

export const CoachMemorySchema = z.object({
  id: z.number(),
  content: z.string(),
  created_on: z.string(),
})

export const CoachSkillSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string(),
})

const PlannedSessionSchema = z.object({
  day: z.string(),
  sport: z.string(),
  workout_type: z.string(),
  description: z.string(),
  target_distance_km: z.number().nullable().optional(),
  target_duration_min: z.number().nullable().optional(),
  intensity: z.string().nullable().optional(),
})

const PlannedWeekSchema = z.object({
  week: z.number(),
  focus: z.string(),
  sessions: z.array(PlannedSessionSchema),
})

export const TrainingPlanSchema = z.object({
  title: z.string(),
  summary: z.string(),
  weeks: z.array(PlannedWeekSchema),
})

export const CoachPlanResponseSchema = z.object({
  plan: TrainingPlanSchema.nullable(),
  message: z.string().nullable(),
})
